Bluetooth: controller: Make ULL/LLL split the default

cvinayak opened the pull request:
Make the Upper and Lower Link Layer split architecture
implementation of the controller as the default when
building Zephyr Bluetooth Low Energy controller support
in applications.

Noticeable missing feature (porting) in comparison to old
architecture implementation is, Advanced scheduling of
connection events.

The missing features will subsequently be submitted
upstream.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Fixes #12681
Depends on #16691

cvinayak committed:
Bluetooth: controller: Make ULL/LLL split the default

Make the Upper and Lower Link Layer split architecture
implementation of the controller as the default when
building Zephyr Bluetooth Low Energy controller support
in applications.

Noticeable missing feature (porting) in comparison to old
architecture implementation is, Advanced scheduling of
connection events.

The missing features will subsequently be submitted
upstream.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jhedberg commented on subsys/bluetooth/controller/ll_sw/ll.c:
```diff
@@ -175,7 +175,10 @@ int ll_init(struct k_sem *sem_rx)
 		return -ENOMEM;
 	}
 
+#if defined(CONFIG_BT_CTLR_FILTER)
```
Possible to use `IS_ENABLED()` here?

cvinayak replied:
I included this change commit from #15780 just to get CI tests to work (as I added new conf files to testcases.yaml). I will remove the commit when #15780 gets merged first.

I kept the #15780 with the commit seperate in order to have backport label on that commit.

aescolar reviewed (dismissed):
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/bluetooth/bsim_bt/bsim_test_app/prj.conf 
was meant to test the legacy controller,
and 
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/bluetooth/bsim_bt/bsim_test_app/prj_split.conf
the split one.
When changing the default, if we want to continue testing the legacy one, 
we would need to be set the 1st one to compile with BT_LL_SW.
(And whenever we are not interested in testing the legacy one anymore, we could just remove it)

carlescufi reviewed (dismissed):
I will block this until we have confirmation that controller-based privacy is not a blocker for anyone, because this is a substantial feature that we are taking away from users.

cvinayak commented:
@aescolar addressed the bsim test conf file in #15849

@aescolar @carlescufi moved commits out into new PR #15849, that are independent from switching the default controller implementation.

tedd-an commented:
I have run PTS with #15849 + #15781 and here is the summary:

Summary:
The test result was same as the previous result which ran without this PRs and **didn't find any other failure**.
[pts_report.xlsx](https://github.com/zephyrproject-rtos/zephyr/files/3143884/pts_report.xlsx)

Some details about the result: The first run was based on the tip of the day + PRs and some GATT TCs failed and it seemed not related to this PR. After rebasing the baseline to the tip used on Wednesday test(and this PR), those failed tests were passed again. I will look into it separately.

sixoctets commented:
@tedd-an Thank you for running the tests. Please upload the .config file of the zephyr build here (just to confirm it built the new link layer with CONFIG_BT_LL_SW_SPLIT=y).

tedd-an commented:
attached .config.
[_config.txt](https://github.com/zephyrproject-rtos/zephyr/files/3148727/_config.txt)

vkchettimada commented:
Thank you

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -16,12 +16,20 @@ if BT_CTLR
 	  Select the Bluetooth Link Layer to compile.
 
+config BT_LL_SW_SPLIT
+	bool "Software-based BLE Link Layer (ULL/LLL split)"
```
Rename this perhaps?
Most people don't know what ULL/LLL split is, or care about that.

cvinayak replied:
@joerchan suggestion?

joerchan replied:
I would suggest to rename this to BT_LL_SW and the old one to BT_LL_SW_LEGACY
Also remove the (ULL/LLL split) part of the description, and put (Legacy) in the description of the old one.

carlescufi replied:
I agree that is the better option too @cvinayak. That way the switch is seamless, although we do slightly change the meaning of a Kconfig option.

aescolar replied:
Just to join the bikeshedding: That would be a bit annoying, as we would need to change the places where today we use BT_LL_SW_SPLIT. Anyhow, chose whichever you guys prefer

cvinayak replied:
How about BT_LL_SW --> BT_LL_SW_LEGACY and keep the BT_LL_SW_SPLIT as is? in a subsequent PR, BT_LL_SW_SPLIT --> BT_LL_SW can be done.

joerchan reviewed (changes requested):
There are also a couple of bugfixes for the legacy controller, that should be ported to the new controller.

cvinayak committed:
Bluetooth: controller: Rename BT_LL_SW to BT_LL_SW_LEGACY

Rename the controller Kconfig option BT_LL_SW to
BT_LL_SW_LEGACY in preparation towards switch to new Link
Layer implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -110,7 +110,7 @@ BLE-enabled builds that can be produced from the Zephyr project codebase:
   * :option:`CONFIG_BT_HCI_RAW` ``=y``
   * :option:`CONFIG_BT_CTLR` ``=y``
-  * :option:`CONFIG_BT_LL_SW` ``=y`` (if using the open source Link Layer)
+  * :option:`CONFIG_BT_LL_SW_LEGACY` ``=y`` (if using the open source Link Layer)
```
fix this to use BT_LL_SW_SPLIT

cvinayak commented on doc/guides/bluetooth/bluetooth-arch.rst:
```diff
@@ -131,7 +131,7 @@ BLE-enabled builds that can be produced from the Zephyr project codebase:
   * :option:`CONFIG_BT_HCI` ``=y``
   * :option:`CONFIG_BT_CTLR` ``=y``
-  * :option:`CONFIG_BT_LL_SW` ``=y`` (if using the open source Link Layer)
+  * :option:`CONFIG_BT_LL_SW_LEGACY` ``=y`` (if using the open source Link Layer)
```
fix this to use BT_LL_SW_SPLIT

cvinayak commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -13,7 +13,7 @@ config COUNTER_NRF_RTC
 	bool "Enable Counter on TIMER0"
 	depends on HAS_HW_NRF_TIMER0
-	depends on !BT_LL_SW
+	depends on !BT_LL_SW_LEGACY
```
Add BT_LL_SW_SPLIT

cvinayak commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -44,7 +44,7 @@ config COUNTER_TIMER4
 	bool "Enable Counter on RTC0"
 	depends on HAS_HW_NRF_RTC0
-	depends on !BT_LL_SW
+	depends on !BT_LL_SW_LEGACY
```
Add BT_LL_SW_SPLIT

dbkinder reviewed (approved):
doc changes LGTM, other than your own comment to``use BT_LL_SW_SPLIT``

cvinayak committed:
Bluetooth: mesh: Use legacy LL for microbit build

Use legacy LL until RAM utilization is optimized in the new
LL.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -16,26 +16,27 @@ if BT_CTLR
 
 choice BT_LL_CHOICE
 	prompt "Bluetooth Link Layer Selection"
-	default BT_LL_SW
```
Why did you remove this line?

cvinayak replied:
BT_LL_SW_SPLIT being the first item in the choice block, it is the de facto default

carlescufi commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -16,26 +16,27 @@ if BT_CTLR
 	  Select the Bluetooth Link Layer to compile.
 
-config BT_LL_SW_LEGACY
+config BT_LL_SW_SPLIT
```
You need to rename this to `BT_LL_SW`, that's what we agreed with @joerchan right? no more "split" in the name

cvinayak replied:
Using the same BT_LL_SW will make users’ config file be valid after switching to new LL but without them noticing the change. Users should know the change in design. I will keep the _SPLIT until legacy is deleted.

carlescufi replied:
That's really not the case if they are just using `CONFIG_BT` on a Nordic board, right? `BT_LL_SW_SPLIT` will be automatically enabled without them noticing

cvinayak replied:
Those who dont use explicit BT_LL_SW dont care if the macro is called BT_LL_SW_SPLIT or anything. We get them to use the new LL. Those who did use (for any reason) BT_LL_SW will now know that it is deprecated and take an informed decision to use BT_LL_SW_SPLIT.
